File: wire-android-automation/common/src/main/java/com/wearezeta/auto/common/email/messages/BackendMessage.java
package com.wearezeta.auto.common.email.messages;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.Map;

import javax.mail.*;
import javax.mail.internet.MimeMessage;

public class BackendMessage {
	private String content;
	private Map<String, String> mapHeaders = new HashMap<String, String>();

	public BackendMessage(String rawMsg) throws Exception {
		final Message msg = stringToMsg(rawMsg);

		@SuppressWarnings("unchecked")
		final Enumeration<Header> hdrs = msg.getAllHeaders();
		while (hdrs.hasMoreElements()) {
			final Header hdr = hdrs.nextElement();
			mapHeaders.put(hdr.getName(), hdr.getValue());
		}

		final Object msgContent = msg.getContent();
		if (msgContent instanceof Multipart) {
			final Multipart multipart = (Multipart) msgContent;
			final StringBuilder multipartContent = new StringBuilder();
			for (int j = 0; j < multipart.getCount(); j++) {
				final BodyPart bodyPart = multipart.getBodyPart(j);
				if (bodyPart.getDisposition() == null) {
					multipartContent.append(getText(bodyPart));
				}
			}
			this.content = multipartContent.toString();
		} else {
			this.content = msgContent.toString();
		}
	}

	public final String getHeaderValue(String headerName) {
		return this.mapHeaders.get(headerName);
	}

	public String getContent() {
		return this.content;
	}

	/**
	 * http://www.oracle.com/technetwork/java/javamail/faq/index.html#mainbody
	 * 
	 * @param p
	 * @return
	 * @throws MessagingException
	 * @throws IOException
	 */
	private String getText(Part p) throws MessagingException, IOException {
		if (p.isMimeType("text/*")) {
			String s = (String) p.getContent();
			p.isMimeType("text/html");
			return s;
		}

		if (p.isMimeType("multipart/alternative")) {
			// prefer html text over plain text
			Multipart mp = (Multipart) p.getContent();
			String text = null;
			for (int i = 0; i < mp.getCount(); i++) {
				Part bp = mp.getBodyPart(i);
				if (bp.isMimeType("text/plain")) {
					if (text == null)
						text = getText(bp);
					continue;
				} else if (bp.isMimeType("text/html")) {
					String s = getText(bp);
					if (s != null)
						return s;
				} else {
					return getText(bp);
				}
			}
			return text;
		} else if (p.isMimeType("multipart/*")) {
			Multipart mp = (Multipart) p.getContent();
			for (int i = 0; i < mp.getCount(); i++) {
				String s = getText(mp.getBodyPart(i));
				if (s != null)
					return s;
			}
		}

		return null;
	}

	public static Message stringToMsg(String rawMsg) throws MessagingException {
		final Session session = Session.getInstance(System.getProperties(), null);
		return new MimeMessage(session, new ByteArrayInputStream(rawMsg.getBytes()));
	}
}

File: tests/testsCore/src/main/kotlin/com/wire/android/tests/core/messages/BackendMessage.kt
/*
 * Wire
 * Copyright (C) 2025 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 */
package com.wire.android.tests.core.messages

import java.io.ByteArrayInputStream
import java.io.IOException
import java.util.Enumeration
import javax.mail.*
import javax.mail.internet.MimeMessage


open class BackendMessage(rawMsg: String) {
    val content: String
    private val mapHeaders = mutableMapOf<String, String>()

    init {
        val msg = stringToMsg(rawMsg)
        val hdrs: Enumeration<Header> = msg.allHeaders as Enumeration<Header>
        while (hdrs.hasMoreElements()) {
            val hdr = hdrs.nextElement()
            mapHeaders[hdr.name] = hdr.value
        }
        content = when (val msgContent = msg.content) {
            is Multipart -> processMultipart(msgContent)
            else -> msgContent.toString()
        }
    }

    fun getHeaderValue(headerName: String): String? = mapHeaders[headerName]

    private fun processMultipart(multipart: Multipart): String {
        var multipartContent = ""
        for (j in 0 until multipart.count) {
            val bodyPart = multipart.getBodyPart(j)
            if (bodyPart.disposition == null) {
                getText(bodyPart)?.let { multipartContent += it }
            }
        }
        return multipartContent
    }

    @Throws(MessagingException::class, IOException::class)
    private fun getText(p: Part): String? {
        if (p.isMimeType("text/*")) {
            return p.content as String
        }

        if (p.isMimeType("multipart/alternative")) {
            // prefer html text over plain text
            val mp = p.content as Multipart
            var text: String? = null
            for (i in 0 until mp.count) {
                val bp = mp.getBodyPart(i)
                when {
                    bp.isMimeType("text/plain") -> {
                        if (text == null) text = getText(bp)
                    }
                    bp.isMimeType("text/html") -> {
                        getText(bp)?.let { return it }
                    }
                    else -> return getText(bp)
                }
            }
            return text
        } else if (p.isMimeType("multipart/*")) {
            val mp = p.content as Multipart
            for (i in 0 until mp.count) {
                getText(mp.getBodyPart(i))?.let { return it }
            }
        }

        return null
    }

    //companion object {
        @Throws(MessagingException::class)
        fun stringToMsg(rawMsg: String): Message {
            val session = Session.getInstance(System.getProperties(), null)
            return MimeMessage(session, ByteArrayInputStream(rawMsg.toByteArray()))
        }
//    }
}
